File: backend/src/main/java/github/lms/lemuel/xr/game/application/ScenarioYamlLoader.java
package github.lms.lemuel.xr.game.application;

import github.lms.lemuel.xr.common.AppException;
import github.lms.lemuel.xr.common.ErrorCode;
import github.lms.lemuel.xr.game.domain.Character;
import github.lms.lemuel.xr.game.domain.Scenario;
import jakarta.annotation.PostConstruct;
import java.io.InputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.EnumMap;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Component;
import org.yaml.snakeyaml.Yaml;

/**
 * resources/scenarios/{character}.yml 로딩.
 *
 * <p>부팅 시 모두 메모리에 로드, runtime 에는 in-memory lookup. 시나리오 변경 시 재배포.
 * Phase 2 에서 hot reload 또는 DB 로 이전.</p>
 */
@Component
@Slf4j
public class ScenarioYamlLoader {

    private final Map<Character, Scenario> cache = new EnumMap<>(Character.class);

    @PostConstruct
    void loadAll() {
        for (Character c : Character.values()) {
            String path = "scenarios/" + c.dbValue() + ".yml";
            try (InputStream in = getClass().getClassLoader().getResourceAsStream(path)) {
                if (in == null) {
                    log.warn("시나리오 yml 없음 — {} 건너뜀: {}", c, path);
                    continue;
                }
                @SuppressWarnings("unchecked")
                Map<String, Object> root = (Map<String, Object>) new Yaml().loadAs(
                        new java.io.InputStreamReader(in, StandardCharsets.UTF_8), Map.class);
                cache.put(c, toScenario(root));
                log.info("시나리오 로드 완료 — {} ({} scenes)", c, cache.get(c).scenes().size());
            } catch (Exception e) {
                log.error("시나리오 로드 실패 — {}: {}", c, e.getMessage(), e);
            }
        }
    }

    public Scenario forCharacter(Character c) {
        Scenario s = cache.get(c);
        if (s == null) throw new AppException(ErrorCode.E_CHARACTER_UNKNOWN, "Scenario not loaded: " + c);
        return s;
    }

    @SuppressWarnings("unchecked")
    private Scenario toScenario(Map<String, Object> root) {
        String character = str(root, "character");
        String title = str(root, "title");
        List<Map<String, Object>> sceneMaps = (List<Map<String, Object>>) root.get("scenes");
        List<Scenario.Scene> scenes = new ArrayList<>();
        for (Map<String, Object> sm : sceneMaps) {
            Map<String, Object> extras = new HashMap<>(sm);
            // 정의된 표준 필드는 extras 에서 제거 → 나머지만 extras 로 보존
            extras.remove("id");
            extras.remove("title");
            extras.remove("type");
            extras.remove("interaction");
            extras.remove("duration_sec");
            extras.remove("narration_id");
            extras.remove("scripture_ref");
            extras.remove("realtime_llm");
            extras.remove("next");
            scenes.add(new Scenario.Scene(
                    (int) sm.get("id"),
                    str(sm, "title"),
                    str(sm, "type"),
                    str(sm, "interaction"),
                    (Integer) sm.get("duration_sec"),
                    str(sm, "narration_id"),
                    str(sm, "scripture_ref"),
                    (Boolean) sm.get("realtime_llm"),
                    (Integer) sm.get("next"),
                    extras
            ));
        }
        return new Scenario(character, title, scenes);
    }

    private static String str(Map<String, Object> m, String k) {
        Object v = m.get(k);
        return v == null ? null : v.toString();
    }
}

File: backend/src/main/kotlin/github/lms/lemuel/xr/game/application/ScenarioYamlLoader.kt
package github.lms.lemuel.xr.game.application

import github.lms.lemuel.xr.common.AppException
import github.lms.lemuel.xr.common.ErrorCode
import github.lms.lemuel.xr.game.domain.Character
import github.lms.lemuel.xr.game.domain.Scenario
import jakarta.annotation.PostConstruct
import org.slf4j.LoggerFactory
import org.springframework.stereotype.Component
import org.yaml.snakeyaml.Yaml
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.util.EnumMap

/**
 * resources/scenarios/{character}.yml 로딩.
 *
 * 부팅 시 모두 메모리에 로드, runtime 에는 in-memory lookup. 시나리오 변경 시 재배포.
 * Phase 2 에서 hot reload 또는 DB 로 이전.
 */
@Component
class ScenarioYamlLoader {

    private val cache: MutableMap<Character, Scenario> = EnumMap(Character::class.java)

    @PostConstruct
    fun loadAll() {
        for (c in Character.entries) {
            val path = "scenarios/${c.dbValue}.yml"
            try {
                javaClass.classLoader.getResourceAsStream(path).use { input ->
                    if (input == null) {
                        log.warn("시나리오 yml 없음 — {} 건너뜀: {}", c, path)
                        return@use
                    }
                    @Suppress("UNCHECKED_CAST")
                    val root = Yaml().loadAs(
                        InputStreamReader(input, StandardCharsets.UTF_8), Map::class.java,
                    ) as Map<String, Any?>
                    cache[c] = toScenario(root)
                    log.info("시나리오 로드 완료 — {} ({} scenes)", c, cache[c]!!.scenes.size)
                }
            } catch (e: Exception) {
                log.error("시나리오 로드 실패 — {}: {}", c, e.message, e)
            }
        }
    }

    fun forCharacter(c: Character): Scenario =
        cache[c] ?: throw AppException(ErrorCode.E_CHARACTER_UNKNOWN, "Scenario not loaded: $c")

    private fun toScenario(root: Map<String, Any?>): Scenario {
        val character = str(root, "character")
        val title = str(root, "title")
        @Suppress("UNCHECKED_CAST")
        val sceneMaps = root["scenes"] as List<Map<String, Any?>>
        val scenes = ArrayList<Scenario.Scene>()
        for (sm in sceneMaps) {
            val extras = HashMap<String, Any?>(sm)
            // 정의된 표준 필드는 extras 에서 제거 → 나머지만 extras 로 보존
            extras.remove("id")
            extras.remove("title")
            extras.remove("type")
            extras.remove("interaction")
            extras.remove("duration_sec")
            extras.remove("narration_id")
            extras.remove("scripture_ref")
            extras.remove("realtime_llm")
            extras.remove("next")
            scenes.add(
                Scenario.Scene(
                    sm["id"] as Int,
                    str(sm, "title"),
                    str(sm, "type"),
                    str(sm, "interaction"),
                    sm["duration_sec"] as Int?,
                    str(sm, "narration_id"),
                    str(sm, "scripture_ref"),
                    sm["realtime_llm"] as Boolean?,
                    sm["next"] as Int?,
                    extras,
                ),
            )
        }
        return Scenario(character, title, scenes)
    }

    companion object {
        private val log = LoggerFactory.getLogger(ScenarioYamlLoader::class.java)

        private fun str(m: Map<String, Any?>, k: String): String? = m[k]?.toString()
    }
}
